src/ocean/task/util/StreamProcessor_test.d

```diff
@@ -21,75 +21,106 @@ import ocean.task.util.Timer;
-    "stream" or "generator" class, one which keeps producing new records for
-    processing at throttled rate
+    Example of processing task, one which is managed by StreamProcessor and
+    gets scheduled for each new record arriving from Generator
 
 *******************************************************************************/
 
-class Generator : Task
+class ProcessingTask : Task
 {
-    void delegate(int) process_dg;
+    int x;
+
+    static int total;
 
-    this ( typeof(this.process_dg) dg )
+    void copyArguments ( int x )
     {
-        this.process_dg = dg;
+        this.x = x;
     }
 
     override void run ( )
     {
-        int i;
-        while (++i)
-            process_dg(i);
+        .wait(100);
+        ++total;
+
+        if (this.x == 1000)
+            theScheduler.shutdown();
     }
 }
 
 /*******************************************************************************
 
-    Example of processing task, one which is managed by StreamProcessor and
-    gets scheduled for each new record arriving from Generator
+    "stream" or "generator" class, one which keeps producing new records for
+    processing at throttled rate
 
 *******************************************************************************/
 
-class ProcessingTask : Task
+class Generator : ISuspendable
 {
-    int x;
+    TimerEvent timer;
+    EpollSelectDispatcher epoll;
+    StreamProcessor!(ProcessingTask) stream_processor;
+    int counter;
 
-    static int total;
+    this ( EpollSelectDispatcher epoll )
+    {
+        this.epoll = epoll;
+        this.timer = new TimerEvent(&this.generate);
 
-    void copyArguments ( int x )
+        auto throttler_config = ThrottlerConfig(10, 1);
+        this.stream_processor = new StreamProcessor!(ProcessingTask)(throttler_config);
+        this.stream_processor.addStream(this);
+    }
+
+    void start ( )
     {
-        this.x = x;
+        this.timer.set(0, 1, 0, 1);
+        this.resume();
     }
 
-    override void run ( )
+    override void resume ( )
     {
-        .wait(100);
-        ++total;
+        this.epoll.register(this.timer);
+    }
 
-        if (this.x == 1000)
-            theScheduler.shutdown();
+    override void suspend ( )
+    {
+        this.epoll.unregister(this.timer);
+    }
+
+    override bool suspended ( )
+    {
+        return this.timer.is_registered;
+    }
+
+    bool generate ( )
+    {
+        this.stream_processor.process(this.counter);
+        ++this.counter;
+        return true;
     }
 }
 
 unittest
 {
+    auto epoll = new EpollSelectDispatcher;
+
     SchedulerConfiguration config;
     config.worker_fiber_limit = 10;
     config.task_queue_limit = 30;
-    initScheduler(config);
-
-    auto throttler_config = ThrottlerConfig(10, 1);
-    auto stream_processor = new StreamProcessor!(ProcessingTask)(throttler_config);
-    auto generator = new Generator(&stream_processor.process);
-    stream_processor.addStream(generator);
+    initScheduler(config, epoll);
```

Comment:
I don't think it is needed to define and pass `epoll` to the scheduler as the scheduler would instantiate it and then `theScheduler.epoll()` can be passed to the Generator. But I might be missing something

Reply:
Correct, it is simply a finger memory from me.